### PR TITLE
Improve CFG for switch expression when clauses

### PR DESCRIPTION
**Pull request by gafter**
Also some related minor changes and improved testing.

### CODE REVIEW

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs [hunk 1]
```diff
@@ -719,6 +719,121 @@ void M(bool result, int input)
...
+Block[B8] - Exit
+    Predecessors: [B7]
+    Statements (0)
+";
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact, WorkItem(32216, "https://github.com/dotnet/roslyn/issues/32216")]
+        public void SwitchExpression_CompoundGuard()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs]: Can you please add some CFG tests for verification of bad switch expr scenarios, or add CFG verification to the existing bad scenario iop tests?

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs [hunk 2]
```diff
@@ -719,6 +719,121 @@ void M(bool result, int input)
...
+        {
+            string source = @"
+#pragma warning disable CS8509
+public sealed class MyClass
+{
+    void M(bool result, int input, bool a, bool b)
+    /*<bind>*/{
+        result = input switch
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs]: >input [](start = 17, length = 5)

Please add some CFG verification for flow in this expression.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs [hunk 3]
```diff
@@ -719,6 +719,121 @@ void M(bool result, int input)
...
+            {
+                1 when a && b => false
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs]: >false [](start = 33, length = 5)

Please add some CFG verification for flow in this expression.

Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchExpression.cs]: >1 [](start = 16, length = 1)

Is it possible for the compiler to bind bad code that has flow in this pattern as a switch expression? If so, please add some CFG verification. If not, could you please add a regular IOp test showing what the behavior is? ie, something like `input switch { test1 && test2 => true }` or similar.

**Comment by 333fred** [comment]
Done review pass (commit 3). Code looks fine, would like some additional tests.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3). Consider adding additional tests suggested by @333fred.

**Pull request merged**